File: .scripts/keychain/provision_telegram_credentials.py
#!/usr/bin/env python3
"""Provision the Telegram api credentials for the signed product binaries.

Owned by BUG-260720-3i74u1; the companion of the control-auth smoke. The
dev provisioning convention (TASK-260716-1iypv4) stores api_id/api_hash in
the login keychain under service `gramdrive-telegram` — but an item
created by the `security` CLI is partition-locked to `apple-tool:`, so the
signed gramdrive-agent reading it hangs on an interactive consent prompt
(fatal for the unattended smoke, and one prompt per fresh item for a
human). This script rewrites the two items so both file-keychain gates
pass silently for the product binaries:

  1. compiles and Developer ID-signs the small Swift provisioning tool
     (`provision-telegram-credentials.swift`) with the same identity that
     signs the app bundle — creating the items from a binary of the
     product's team puts the *partition list* on `teamid:<team>`;
  2. runs it with a trusted-application ACL naming the packaged
     gramdrive-agent and GramDrive binaries — covering the *ACL* gate.

Values come from GRAMDRIVE_API_ID/GRAMDRIVE_API_HASH env when set,
otherwise they are captured from the existing keychain items via the
`security` CLI (which may itself prompt once if the items were already
rewritten — pass the env to stay unattended). Secrets never touch argv,
logs, or the repo.

Usage: python3 .scripts/keychain/provision_telegram_credentials.py
           [--identity 'Developer ID Application: ...']
           [--bundle .temp/app-packaging/GramDrive.app]
"""

from __future__ import annotations

import argparse
import os
import re
import subprocess
import sys
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[2]
TOOL_SOURCE = Path(__file__).with_name("provision-telegram-credentials.swift")
BUILD_DIR = REPO_ROOT / ".temp" / "keychain-provision"
SERVICE = "gramdrive-telegram"


def fail(message: str) -> "int":
    print(f"provision: {message}", file=sys.stderr)
    return 1


def resolve_identity(explicit: str | None) -> str | None:
    if explicit:
        return explicit
    listing = subprocess.run(
        ["security", "find-identity", "-v", "-p", "codesigning"],
        capture_output=True, text=True, check=False,
    ).stdout
    match = re.search(r'"(Developer ID Application: [^"]+)"', listing)
    return match.group(1) if match else None


def capture_existing(account: str) -> str | None:
    probe = subprocess.run(
        ["security", "find-generic-password", "-s", SERVICE, "-a", account, "-w"],
        capture_output=True, text=True, check=False,
    )
    value = probe.stdout.strip()
    return value if probe.returncode == 0 and value else None


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--identity", help="Developer ID Application identity name")
    parser.add_argument(
        "--bundle",
        type=Path,
        default=REPO_ROOT / ".temp" / "app-packaging" / "GramDrive.app",
        help="packaged app bundle whose binaries get keychain access",
    )
    args = parser.parse_args()

    agent = args.bundle / "Contents" / "MacOS" / "gramdrive-agent"
    app = args.bundle / "Contents" / "MacOS" / "GramDrive"
    if not agent.is_file() or not app.is_file():
        return fail(f"packaged bundle not found at {args.bundle}; run `make package-app`")

    identity = resolve_identity(args.identity)
    if not identity:
        return fail("no Developer ID Application identity in the keychain")

    api_id = os.environ.get("GRAMDRIVE_API_ID") or capture_existing("api_id")
    api_hash = os.environ.get("GRAMDRIVE_API_HASH") or capture_existing("api_hash")
    if not api_id or not api_hash:
        return fail(
            "api credentials unavailable: set GRAMDRIVE_API_ID/GRAMDRIVE_API_HASH "
            f"or provision the `{SERVICE}` keychain items first"
        )

    BUILD_DIR.mkdir(parents=True, exist_ok=True)
    tool = BUILD_DIR / "provision-telegram-credentials"
    compile_run = subprocess.run(
        ["swiftc", "-O", "-o", str(tool), str(TOOL_SOURCE)],
        capture_output=True, text=True, check=False,
    )
    if compile_run.returncode != 0:
        return fail(f"tool compile failed:\n{compile_run.stderr}")
    sign_run = subprocess.run(
        ["codesign", "--force", "--sign", identity, "--timestamp", str(tool)],
        capture_output=True, text=True, check=False,
    )
    if sign_run.returncode != 0:
        return fail(f"tool signing failed:\n{sign_run.stderr}")

    # Prove the freshly signed tool actually loads and runs on THIS machine
    # (Gatekeeper, quarantine, a bad link) before we clear the existing
    # keychain items — a tool that cannot run must not first destroy the only
    # copies of the credentials.
    check_run = subprocess.run(
        [str(tool), "--check"], capture_output=True, text=True, check=False,
    )
    if check_run.returncode != 0:
        return fail(f"provisioning tool is not runnable:\n{check_run.stderr}")

    # Items created by the `security` CLI refuse a delete from the signed
    # tool (errSecInvalidOwnerEdit); their creator removes them. Items the
    # tool itself created earlier are replaced by the tool.
    for account in ("api_id", "api_hash"):
        subprocess.run(
            ["security", "delete-generic-password", "-s", SERVICE, "-a", account],
            capture_output=True, check=False,
        )

    environment = dict(os.environ)
    environment["GRAMDRIVE_API_ID"] = api_id
    environment["GRAMDRIVE_API_HASH"] = api_hash
    provision = subprocess.run(
        [str(tool), "--agent", str(agent), "--app", str(app)],
        env=environment, check=False,
    )
    if provision.returncode != 0:
        return fail("provisioning tool failed")
    print(f"provisioned: service `{SERVICE}` readable by {agent.name} and {app.name}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .scripts/keychain/provision-telegram-credentials.swift
// Provision the Telegram api credentials into the login keychain so the
// *signed product binaries* can read them without a consent prompt
// (BUG-260720-3i74u1).
//
// Why a signed Swift tool and not `security add-generic-password`: a file
// keychain item carries two independent gates — the trusted-application
// ACL and the partition list (the creator's code-signing partition,
// e.g. `apple-tool:` for the `security` CLI, `teamid:...` for a signed
// app). An item created by the `security` CLI is partition-locked to
// `apple-tool:`, so a Developer ID binary reading it always triggers the
// interactive consent prompt, ACL or not — and changing the partition
// list of an existing item requires the keychain password. Creating the
// items *from a binary signed with the product's team* sets the partition
// to that team at creation, and an explicit trusted-application ACL
// covers the product binaries; both gates then pass silently.
//
// Usage (driven by provision_telegram_credentials.py, which compiles and
// signs this source with the product's Developer ID identity):
//
//   provision-telegram-credentials --agent PATH --app PATH
//   provision-telegram-credentials --check   # prove the binary runs; no I/O
//
// The api_id/api_hash values are read from the GRAMDRIVE_API_ID /
// GRAMDRIVE_API_HASH environment variables — never from argv (visible in
// `ps`) and never echoed.
//
// The tool only ever *writes* the credentials (from its environment); it has
// no read/consume path. A signed binary that could read these items and hand
// them to an arbitrary command would be a promptless secret-exfiltration
// primitive, so no such mode exists and the tool does not name itself in the
// items' ACL — only the product binaries (and the `security` CLI for dev
// inspection) may read them.

import Foundation
import Security

let service = "gramdrive-telegram"

func fail(_ message: String) -> Never {
    FileHandle.standardError.write(Data(("provision: " + message + "\n").utf8))
    exit(1)
}

var arguments = Array(CommandLine.arguments.dropFirst())
// A side-effect-free liveness probe: the driver runs it after signing to
// prove the binary loads and runs on this machine *before* it clears the
// existing keychain items. Reads and writes nothing.
if arguments.first == "--check" {
    print("provision-telegram-credentials: ok")
    exit(0)
}
// The trusted-application ACL never includes this tool: it has no read path,
// so self-trust would only widen the items' reach for no benefit.
var trustedPaths: [String] = ["/usr/bin/security"]
while !arguments.isEmpty {
    let flag = arguments.removeFirst()
    guard flag == "--agent" || flag == "--app", !arguments.isEmpty else {
        fail("usage: provision-telegram-credentials --agent PATH --app PATH")
    }
    trustedPaths.append(arguments.removeFirst())
}

let environment = ProcessInfo.processInfo.environment
guard
    let apiId = environment["GRAMDRIVE_API_ID"], !apiId.isEmpty,
    let apiHash = environment["GRAMDRIVE_API_HASH"], !apiHash.isEmpty
else {
    fail("GRAMDRIVE_API_ID / GRAMDRIVE_API_HASH must be set in the environment")
}

// The trusted-application ACL: the product binaries and the security CLI
// (dev inspection) only — never this tool, which has no read path. The
// SecTrustedApplication/SecAccess API family is deprecated but remains
// the only way to author file-keychain ACLs, which is exactly what the
// product's vault reads (SecItemCopyMatching over the login keychain).
var trusted: [SecTrustedApplication] = []
for path in trustedPaths {
    var application: SecTrustedApplication?
    let status = SecTrustedApplicationCreateFromPath(path, &application)
    guard status == errSecSuccess, let application else {
        fail("cannot trust \(path): \(status)")
    }
    trusted.append(application)
}
var access: SecAccess?
let accessStatus = SecAccessCreate("Telegram api credentials" as CFString, trusted as CFArray, &access)
guard accessStatus == errSecSuccess, let access else {
    fail("SecAccessCreate failed: \(accessStatus)")
}

for (account, value) in [("api_id", apiId), ("api_hash", apiHash)] {
    let match: [String: Any] = [
        kSecClass as String: kSecClassGenericPassword,
        kSecAttrService as String: service,
        kSecAttrAccount as String: account,
    ]
    // Replacement of an item this tool created earlier; items created by
    // other owners (the `security` CLI) refuse the delete with
    // errSecInvalidOwnerEdit and are removed by the driver script instead —
    // if one survives anyway, the add below reports the duplicate.
    let deleteStatus = SecItemDelete(match as CFDictionary)
    guard
        deleteStatus == errSecSuccess || deleteStatus == errSecItemNotFound
            || deleteStatus == errSecInvalidOwnerEdit
    else {
        fail("cannot replace \(account): delete failed \(deleteStatus)")
    }
    var add = match
    add[kSecValueData as String] = Data(value.utf8)
    add[kSecAttrAccess as String] = access
    let addStatus = SecItemAdd(add as CFDictionary, nil)
    guard addStatus == errSecSuccess else {
        fail("cannot add \(account): \(addStatus)")
    }
    print("provisioned \(service)/\(account)")
}
